# Employment in AI, Data Science and Software Development (USA, May 2023)

**Module:** Data Visualisation and Communication – CA1
**Student:** Silmara Elisangela Souto – 2023357

**GitHub repository:** https://github.com/2023357/data-vis-ca1
**Video presentation:** 


## Introduction

Artificial Intelligence (AI) is one of the fastest-growing areas of the technology sector,
and it depends on professionals with skills in software development, data science and computer research.
This notebook uses data visualisation to explore employment in these fields in the United States.

The analysis uses the Occupational Employment and Wage Statistics (OEWS) dataset for May 2023, published by the U.S. Bureau of Labor Statistics (BLS). 
Because the BLS does not have a specific "AI" occupation category, this analysis selects a group of occupations closely related to AI,
such as Software Developers, Data Scientists and Computer and Information Research Scientists.

The notebook aims to answer the following questions:
1. Which AI-related occupations employ the most people?
2. How do wages compare across these occupations?
3. How do employment and wages vary across US states?
4. How has employment in AI-related occupations changed over time?

For each visualisation, the notebook explains the objective, why that type of chart was chosen, and what it reveals about AI employment.

## 1. Loading the Data.

In [6]:
# Import libraries for data handling and visualisation

# Load the BLS OEWS May 2023 dataset (this may take a few minutes)

# Display the first five rows to check the data loaded correctly


import pandas as pd
import seaborn as sns 
import matplotlib.pyplot as plt

oes_df = pd.read_excel("oesm23all/all_data_M_2023.xlsx")

oes_df.head()





,AREA,AREA_TITLE,AREA_TYPE,PRIM_STATE,NAICS,NAICS_TITLE,I_GROUP,OWN_CODE,OCC_CODE,OCC_TITLE,...,H_MEDIAN,H_PCT75,H_PCT90,A_PCT10,A_PCT25,A_MEDIAN,A_PCT75,A_PCT90,ANNUAL,HOURLY
0,99,U.S.,1,US,000000,Cross-industry,cross-industry,1235,00-0000,All Occupations,...,23.11,37.01,58.4,29050,35660,48060,76980,121470,NaN,NaN
1,99,U.S.,1,US,000000,Cross-industry,cross-industry,1235,11-0000,Management Occupations,...,56.19,81.29,111.36,54550,78330,116880,169090,231620,NaN,NaN
2,99,U.S.,1,US,000000,Cross-industry,cross-industry,1235,11-1000,Top Executives,...,49.74,79.57,#,46400,66170,103460,165500,#,NaN,NaN
3,99,U.S.,1,US,000000,Cross-industry,cross-industry,1235,11-1010,Chief Executives,...,99.37,#,#,80000,130840,206680,#,#,NaN,NaN
4,99,U.S.,1,US,000000,Cross-industry,cross-industry,1235,11-1011,Chief Executives,...,99.37,#,#,80000,130840,206680,#,#,NaN,NaN


In [10]:
# Check the number of rows and columns in the dataset
oes_df.shape

(413327, 32)

In [11]:

oes_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 413327 entries, 0 to 413326
Data columns (total 32 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   AREA          413327 non-null  int64 
 1   AREA_TITLE    413327 non-null  object
 2   AREA_TYPE     413327 non-null  int64 
 3   PRIM_STATE    413327 non-null  object
 4   NAICS         413327 non-null  object
 5   NAICS_TITLE   413327 non-null  object
 6   I_GROUP       413327 non-null  object
 7   OWN_CODE      413327 non-null  int64 
 8   OCC_CODE      413327 non-null  object
 9   OCC_TITLE     413327 non-null  object
 10  O_GROUP       413327 non-null  object
 11  TOT_EMP       413327 non-null  object
 12  EMP_PRSE      413327 non-null  object
 13  JOBS_1000     235826 non-null  object
 14  LOC_QUOTIENT  235826 non-null  object
 15  PCT_TOTAL     170470 non-null  object
 16  PCT_RPT       170470 non-null  object
 17  H_MEAN        413327 non-null  object
 18  A_MEAN        413327 non

## 2. Understanding the Data

The dataset comes from the U.S. Bureau of Labor Statistics (BLS) Occupational Employment and Wage Statistics (OEWS) survey and contains employment and wage estimates for May 2023. It has 413,327 rows and 32 columns.

Each row represents one occupation in a specific geographic area (the whole country, a state or a metropolitan area) and, in some cases, a specific industry. Because of this, the same occupation appears many times in the dataset, and the data must be filtered to the right level before it can be analysed.

The most relevant columns for this analysis are:
- `OCC_TITLE`: the name of the occupation
- `TOT_EMP`: the estimated number of people employed
- `A_MEAN` and `A_MEDIAN`: the mean and median annual wage (USD)
- `AREA_TITLE`: the geographic area

The `info()` output shows that numeric columns such as `TOT_EMP` and `A_MEAN` were loaded as `object` (text) instead of numbers. This happens because the BLS uses special symbols in place of some values: `*` and `**` when an estimate is not available, and `#` when a wage is above the highest value the BLS publishes. These columns therefore need to be cleaned and converted to numbers before any visualisation can be created.

Only three columns (`AREA`, `AREA_TYPE` and `OWN_CODE`) were loaded as integers. Some columns, such as `JOBS_1000` and `LOC_QUOTIENT`, also contain missing values, because they are only available for state and metropolitan area estimates.

## 3. Data Cleaning

In [12]:
(oes_df['TOT_EMP'] == '**').sum()

np.int64(11598)

In [16]:
(oes_df['A_MEAN'] == '*').sum()


np.int64(5083)

In [17]:
(oes_df['A_MEAN'] == '#').sum()

np.int64(269)

In [20]:
oes_df['TOT_EMP'] = oes_df['TOT_EMP'] = pd.to_numeric(oes_df['TOT_EMP'], errors="coerce")
oes_df['A_MEAN'] = pd.to_numeric(oes_df['A_MEAN'], errors="coerce")
oes_df['A_MEDIAN'] = pd.to_numeric(oes_df['A_MEDIAN'], errors="coerce")
oes_df['A_PCT10'] = pd.to_numeric(oes_df['A_PCT10'], errors="coerce")
oes_df['A_PCT25'] = pd.to_numeric(oes_df['A_PCT25'], errors="coerce")
oes_df['A_PCT75'] = pd.to_numeric(oes_df['A_PCT75'], errors="coerce")
oes_df['A_PCT90'] = pd.to_numeric(oes_df['A_PCT90'], errors="coerce")
oes_df['JOBS_1000'] = pd.to_numeric(oes_df['JOBS_1000'], errors="coerce")
oes_df['LOC_QUOTIENT'] = pd.to_numeric(oes_df['LOC_QUOTIENT'], errors="coerce")

In [21]:
oes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 413327 entries, 0 to 413326
Data columns (total 32 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   AREA          413327 non-null  int64  
 1   AREA_TITLE    413327 non-null  object 
 2   AREA_TYPE     413327 non-null  int64  
 3   PRIM_STATE    413327 non-null  object 
 4   NAICS         413327 non-null  object 
 5   NAICS_TITLE   413327 non-null  object 
 6   I_GROUP       413327 non-null  object 
 7   OWN_CODE      413327 non-null  int64  
 8   OCC_CODE      413327 non-null  object 
 9   OCC_TITLE     413327 non-null  object 
 10  O_GROUP       413327 non-null  object 
 11  TOT_EMP       401729 non-null  float64
 12  EMP_PRSE      413327 non-null  object 
 13  JOBS_1000     232193 non-null  float64
 14  LOC_QUOTIENT  232193 non-null  float64
 15  PCT_TOTAL     170470 non-null  object 
 16  PCT_RPT       170470 non-null  object 
 17  H_MEAN        413327 non-null  object 
 18  A_ME

## Handling special symbols
As shown in the previous section, the numeric columns were loaded as text because the BLS uses special symbols in place of some values:
- `*` indicates that a wage estimate is not available
- `**` indicates that an employment estimate is not available
- `#` indicates a wage equal to or greater than $239,200 per year, the highest value the BLS publishes

For example, 11,598 rows of `TOT_EMP` contained `**`, and 5,083 rows of `A_MEAN` contained `*`.

The nine columns used in this analysis were converted to numbers using `pd.to_numeric` with `errors="coerce"`, which replaces any value that cannot be converted with `NaN` (missing value). After the conversion, the `info()` output shows these columns as `float64`.

**Handling the `#` values**

Unlike `*` and `**`, the `#` symbol does not represent missing data. It means the wage is top-coded: the real value exists but is higher than the maximum the BLS publishes ($239,200 per year). Three options were considered:

1. **Convert `#` to `NaN`** – simple and transparent, as no values are invented, but it slightly underestimates the highest wages.
2. **Replace `#` with $239,200** – keeps the information that the wage is high, but assigns a value that is likely lower than the real one.
3. **Remove the rows** – rejected, because it would also remove valid employment and wage values from the same rows.

Option 1 was chosen. A check of the AI-related occupations selected for this project showed that `#` does not appear in their mean or median wages, and appears only five times in the state-level 90th percentile wage (`A_PCT90`). Since the visualisations focus on mean and median wages, the impact of this decision on the analysis is minimal. Option 2 would be more appropriate in an analysis focused on the highest wages.



## 4. Selecting AI-related Occupations